packages/docusaurus-mdx-loader/src/remark/toc/index.ts

```diff
@@ -93,21 +93,75 @@ const plugin: Plugin = function plugin(
     const {toString} = await import('mdast-util-to-string');
     const {visit} = await import('unist-util-visit');
 
-    const headings: TOCItem[] = [];
+    const partialComponentToHeadingsName: {[key: string]: string} =
+      Object.create(null);
 
-    visit(root, 'heading', (child: Heading) => {
-      const value = toString(child);
+    // TOCItem or string already with the spread operator
+    const headings: (TOCItem | string)[] = [];
 
-      // depth:1 headings are titles and not included in the TOC
-      if (!value || child.depth < 2) {
-        return;
+    visit(root, ['heading', 'mdxjsEsm', 'mdxJsxFlowElement'], (child) => {
+      if (child.type === 'heading') {
+        const headingNode = child as Heading;
+        const value = toString(headingNode);
+
+        // depth:1 headings are titles and not included in the TOC
+        if (!value || headingNode.depth < 2) {
+          return;
+        }
+
+        headings.push({
+          value: toValue(headingNode, toString),
+          id: headingNode.data!.id!,
+          level: headingNode.depth,
+        });
       }
 
-      headings.push({
-        value: toValue(child, toString),
-        id: child.data!.id!,
-        level: child.depth,
-      });
+      if (child.type === 'mdxjsEsm') {
```

Comment:
I know the original code was already low-level, but I'd appreciate if we could make it more readable and high-level. I mean the intent of this algorithm should be clear without spending time trying to understand the implementation details. Currently it's a bit hard to see the big picture of it. I would extract things in smaller functions for example.

Reply:
Moved the visitor for each type of node into its own function. This is a bit more readable now, but I can try to do some further reorganisation if needed.